File: src/cloudmorrow/tui/panes/kit_grouped.py
"""A `list` picked through, in the terminal: groups down the left, a subgroup bar, the rows.

The kit's list (panes/kit.py) with a `group` and perhaps a `subgroup` bound
(docs/QUILLS.md). Laid out as notes are: the group's values in a list you
pick from on the left, and beside it the subgroup's values as a row of
buttons, and under them the records in both. A new record goes into the
group and subgroup on screen.

The values of a level are the records a link field points at, an enum's
values, or — for an indexed string — the values the records have, with a
way to name a new one (`n` in the list, `e` for the subgroup). A value just
named is a row before anything is in it, or there would be nowhere to put
the first record.

A field the datamodel marks `secret` is never in a listing — the server
sends null — so its column says there is something there, in dots, until
`v` fetches that one record and shows it; `v` again hides it. `c` copies it
without showing it. The record sheet draws it as a password box.

Nothing here knows what it is drawing: a Secrets vault and environment are
a group and a subgroup, as a customer and a project would be.
"""

from __future__ import annotations

from textual import work
from textual.app import ComposeResult
from textual.containers import Horizontal, Vertical
from textual.widgets import Button, DataTable, Static

from cloudmorrow.client.api import ApiError
from cloudmorrow.tui.panes.kit import KitPane, ListPane
from cloudmorrow.tui.screens.modals import PromptModal
from cloudmorrow.tui.screens.record_sheet import shown
from cloudmorrow.tui.theme import MUTED
from cloudmorrow.tui.widgets.group_list import GroupList
from cloudmorrow.tui.widgets.kit import field_label, field_of, safe_id, title_of
from cloudmorrow.tui.widgets.toolbar import Action

MASK = "••••••••"


def draws_here(model: dict, screen: dict) -> bool:
    """A list screen this pane draws rather than the plain one: grouped, or hiding a field."""
    subtitle = field_of(model, screen.get("subtitle"))
    return bool(screen.get("group") or (subtitle and subtitle.get("secret")))


class SubgroupBar(Horizontal):
    """The subgroup's values, as buttons. Clicking one stands on it."""

    def __init__(self, **kwargs) -> None:
        super().__init__(**kwargs)
        self.add_class("kit-subgroup-bar")
        self.values: dict[str, str] = {}

    async def show(self, choices: list[tuple[str, str, int]], current: str | None, noun: str) -> None:
        # Awaited: removal is scheduled, and remounting the same ids before it
        # lands is a duplicate-id error.
        await self.remove_children()
        self.values = {}
        for value, label, count in choices:
            wid = f"sub-{safe_id(value)}"
            self.values[wid] = value
            button = Button(
                f"{label}  {count}" if count else label,
                id=wid,
                compact=True,
                variant="primary" if value == current else "default",
            )
            button.tooltip = f"{count} in {label}"
            await self.mount(button)
        add = Button("＋", id="sub-new", compact=True)
        add.tooltip = f"Another {noun}"
        await self.mount(add)


class GroupedListPane(ListPane):
    """`list` with a group (and a subgroup), or with a secret field, opened in the sheet."""

    BINDINGS = [
        *KitPane.BINDINGS,
        ("v", "fire('reveal')", "Reveal"),
        ("c", "fire('copy')", "Copy"),
        ("e", "fire('new_subgroup')", "New subgroup"),
        ("r", "fire('refresh')", "Refresh"),
    ]

    def __init__(self, quill: dict, screen: dict, **kwargs) -> None:
        super().__init__(quill, screen, **kwargs)
        self.levels: list[dict] = [
            f for f in (field_of(self.model, screen.get(n)) for n in ("group", "subgroup")) if f
        ]
        self.hidden = self.subtitle if self.subtitle and self.subtitle.get("secret") else None
        self.all: list[dict] = []
        self.chosen: list[str | None] = [None] * len(self.levels)
        self.offered: list[list[tuple[str, str, int]]] = [[] for _ in self.levels]
        # Values fetched by `v`, by record id, while this pane is up.
        self.revealed: dict[str, str] = {}
        actions = [Action("new_record", f"New {self.noun}", "^n", variant="primary")]
        actions.append(Action("open_record", "Open", "enter"))
        if self.hidden:
            label = field_label(self.hidden).lower()
            actions.append(Action("reveal", "Reveal", "v", hint=f"Show the {label}, or hide it again"))
            actions.append(Action("copy", "Copy", "c", hint=f"Copy the {label} without showing it"))
        if self.tick:
            actions.append(Action("tick", field_label(self.tick), "space"))
        actions.append(Action("delete_record", "Delete", "del", variant="error"))
        self.ACTIONS = tuple(actions)

    def content(self) -> ComposeResult:
        with Horizontal(classes="kit-grouped"):
            if self.levels:
                yield GroupList(id="kit-group-list", classes="kit-group-list")
            with Vertical(classes="kit-group-side"):
                if self.levels:
                    yield Static("", id="kit-group-name", classes="kit-group-name")
                if len(self.levels) > 1:
                    yield SubgroupBar(id="kit-subgroup-bar")
                with Vertical(classes="kit-group-rows"):
                    yield DataTable(id="kit-table", cursor_type="row", zebra_stripes=True)
                    yield Static("", id="kit-group-empty", classes="kit-group-empty")

    def on_mount(self, event) -> None:
        # The plain list's columns are not these: its handler is not run too.
        event.prevent_default()
        table = self.query_one("#kit-table", DataTable)
        columns = [" "] if self.tick else []
        title = field_of(self.model, self.title_field)
        columns.append(field_label(title).upper() if title else "")
        if self.subtitle:
            columns.append(field_label(self.subtitle).upper())
        columns.append("UPDATED")
        table.add_columns(*columns)

    def on_show(self) -> None:
        self.reload()
        # Focus lands in the pane so its keys are in the footer before you
        # touch anything, unless you are already working inside it.
        focused = self.screen.focused
        if focused is None or self not in focused.ancestors:
            target = self.query("#kit-group-list") or self.query("#kit-table")
            target.first().focus()

    # -- the levels ----------------------------------------------------------
    async def _choices(self, field: dict, records: list[dict]) -> list[tuple[str, str]]:
        if field.get("kind") == "link":
            target = self.models.get(field.get("to")) or {}
            try:
                rows = await self.api.records(field["to"])
            except ApiError:
                rows = []
            return [(str(r["id"]), title_of(r, target)) for r in rows]
        if field.get("kind") == "enum":
            labels = field.get("labels") or field.get("values") or []
            return list(zip(field.get("values") or [], labels, strict=False))
        seen = {r["fields"].get(field["name"]) for r in records}
        return [(str(v), str(v)) for v in sorted(v for v in seen if v not in (None, ""))]

    async def _place(self) -> None:
        """Work out each level's choices and where it stands, then draw it all."""
        inside = self.all
        for index, field in enumerate(self.levels):
            choices = await self._choices(field, inside)
            values = [value for value, _ in choices]
            current = self.chosen[index]
            # Where it stood stays where it stands, even emptied or just named.
            if current is None:
                default = field.get("default")
                current = (
                    str(default) if default is not None and str(default) in values
                    else values[0] if values else (str(default) if default is not None else None)
                )
            if current is not None and current not in values:
                choices.append((current, current))
            self.chosen[index] = current
            counted = [
                (value, label, sum(1 for r in inside if str(r["fields"].get(field["name"])) == value))
                for value, label in choices
            ]
            self.offered[index] = counted
            inside = [r for r in inside if current is None or str(r["fields"].get(field["name"])) == current]
        self.records = inside
        await self._draw_levels()
        self.draw()

    async def _draw_levels(self) -> None:
        if not self.levels:
            return
        self.query_one(GroupList).show(self.offered[0], self.chosen[0])
        names = [self._label_of(i) for i in range(len(self.levels)) if self.chosen[i] is not None]
        self.query_one("#kit-group-name", Static).update(f"[b]{' · '.join(names)}[/]")
        if len(self.levels) > 1:
            await self.query_one(SubgroupBar).show(
                self.offered[1], self.chosen[1], field_label(self.levels[1]).lower()
            )

    def _label_of(self, index: int) -> str:
        value = self.chosen[index]
        return next((label for v, label, _ in self.offered[index] if v == value), str(value))

    # -- loading -----------------------------------------------------------
    @work(exclusive=True, group="kit-list")
    async def reload(self, message: str = "", keep: str | None = None) -> None:
        if self.api is None:
            return
        try:
            self.all = await self.api.records(self.model_id)
        except ApiError as exc:
            await self.signed_out(exc)
            return
        self.loaded = True
        if keep is not None:
            # Stand where the record just saved is, whatever it was named as.
            record = next((r for r in self.all if r["id"] == keep), None)
            if record is not None:
                self.chosen = [str(record["fields"].get(f["name"])) for f in self.levels]
        await self._place()
        if keep is not None:
            self.draw(keep=keep)
        if message:
            self.status(message)

    def draw(self, *, keep: str | None = None) -> None:
        super().draw(keep=keep)
        empty = self.query_one("#kit-group-empty", Static)
        table = self.query_one("#kit-table", DataTable)
        table.display = bool(self.records)
        empty.display = not self.records
        if not self.records:
            where = " · ".join(self._label_of(i) for i in range(len(self.levels)) if self.chosen[i])
            place = f" in [/][b]{where}[/][{MUTED}]" if where else ""
            empty.update(f"[{MUTED}]Nothing{place} yet. New {self.noun} (^n) starts one here.[/]")
            self.status("")

    def _row(self, record: dict) -> tuple[str, ...]:
        cells = list(super()._row(record))
        if self.hidden:
            value = self.revealed.get(record["id"])
            cells[-1] = f"[{MUTED}]{MASK}[/]" if value is None else (value or f"[{MUTED}](empty)[/]")
        stamp = str(record.get("updated_at") or "")
        cells.append(f"[{MUTED}]{shown({'kind': 'datetime'}, stamp) if stamp else ''}[/]")
        return tuple(cells)

    def card_status(self) -> tuple[str, str] | None:
        if not self.loaded:
            return None
        count = len(self.all)
        said = f"{count} {self.noun}{'' if count == 1 else 's'}"
        if self.levels:
            groups = len([g for g in self.offered[0] if g[2]])
            noun = field_label(self.levels[0]).lower()
            said = f"{groups} {noun}{'' if groups == 1 else 's'}, {said}"
        return "ok", said

    # -- picking -------------------------------------------------------------
    def on_group_list_chosen(self, event: GroupList.Chosen) -> None:
        event.stop()
        if event.value != self.chosen[0]:
            self.stand(0, event.value)

    def on_group_list_new_requested(self, _: GroupList.NewRequested) -> None:
        self.name_value(0)

    def on_button_pressed(self, event: Button.Pressed) -> None:
        wid = event.button.id or ""
        if not wid.startswith("sub-"):
            return
        event.stop()
        if wid == "sub-new":
            self.name_value(1)
            return
        bar = self.query_one(SubgroupBar)
        if wid in bar.values:
            self.stand(1, bar.values[wid])

    @work(exclusive=True, group="kit-list")
    async def stand(self, index: int, value: str) -> None:
        """Stand on another value of a level; the levels under it start again."""
        self.chosen[index] = value
        for below in range(index + 1, len(self.levels)):
            self.chosen[below] = None
        # What was on screen belongs to where it was revealed.
        self.revealed.clear()
        await self._place()

    def act_new_group(self) -> None:
        self.name_value(0)

    def act_new_subgroup(self) -> None:
        if len(self.levels) > 1:
            self.name_value(1)

    @work(group="ui")
    async def name_value(self, index: int) -> None:
        if index >= len(self.levels):
            return
        field = self.levels[index]
        noun = field_label(field).lower()
        if field.get("kind") == "enum":
            self.status(f"a {noun} is one of the ones there are", error=True)
            return
        text = await self.app.push_screen_wait(
            PromptModal(
                f"New {noun}",
                placeholder=str(field.get("default") or ""),
                detail=f"[dim]A {noun} is there once something is in it: add one next.[/]",
            )
        )
        text = (text or "").strip()
        if not text:
            return
        value = text
        if field.get("kind") == "link":
            target = self.models.get(field["to"]) or {}
            try:
                made = await self.api.create_record(field["to"], {target.get("title", "title"): text})
            except ApiError as exc:
                await self.signed_out(exc)
                return
            value = str(made["id"])
        self.stand(index, value)

    # -- the records -----------------------------------------------------------
    @work(group="ui")
    async def new_record(self) -> None:
        preset = {f["name"]: v for f, v in zip(self.levels, self.chosen, strict=False) if v is not None}
        if self.tick:
            preset[self.tick["name"]] = False
        saved = await self.open_sheet(None, preset=preset)
        if not isinstance(saved, dict):
            return
        self.reload(f"Added {title_of(saved, self.model)}.", keep=saved["id"])

    @work(group="ui")
    async def edit_record(self, record: dict) -> None:
        result = await self.open_sheet(record)
        if result is None:
            return
        if result == "deleted":
            self.revealed.pop(record["id"], None)
            self.reload(f"Deleted {title_of(record, self.model)}.")
            return
        self.revealed.pop(record["id"], None)
        self.reload(keep=result["id"] if isinstance(result, dict) else None)

    @work(group="ui")
    async def delete_record(self, record: dict) -> None:
        if await self.confirm_delete(record):
            self.revealed.pop(record["id"], None)
            self.reload(f"Deleted {title_of(record, self.model)}.")

    def act_refresh(self) -> None:
        self.reload()

    def act_reveal(self) -> None:
        self.reveal()

    @work(group="ui")
    async def reveal(self) -> None:
        record = self.selected
        if record is None or self.hidden is None:
            return
        name = title_of(record, self.model)
        if record["id"] in self.revealed:
            del self.revealed[record["id"]]
            self.draw(keep=record["id"])
            return
        value = await self._value_of(record)
        if value is None:
            return
        self.revealed[record["id"]] = value
        self.draw(keep=record["id"])
        self.status(f"{name} is on screen — reveal again to hide it.")

    def act_copy(self) -> None:
        self.copy_value()

    @work(group="ui")
    async def copy_value(self) -> None:
        record = self.selected
        if record is None or self.hidden is None:
            return
        value = self.revealed.get(record["id"])
        if value is None:
            value = await self._value_of(record)
        if value is None:
            return
        # Textual copies via OSC 52, so this works over SSH in most terminals.
        self.app.copy_to_clipboard(value)
        self.status(f"Copied {title_of(record, self.model)} to the clipboard.")

    async def _value_of(self, record: dict) -> str | None:
        """The hidden field of one record: asked for, one record at a time."""
        try:
            full = await self.api.record(self.model_id, record["id"])
        except ApiError as exc:
            await self.signed_out(exc)
            return None
        return str((full.get("fields") or {}).get(self.hidden["name"]) or "")

File: src/cloudmorrow/server/web/kit_grouped.js
/* The kit's list, picked through: a `list` screen with a `group`, and
   perhaps a `subgroup` under it, and the kit's way with a secret field.

   A group is a row of chips across the top, one per value of its field:
   the records a link points at, an enum's values, or — for an indexed
   string — the values the records have, with "+ New" on the end to name
   another. A subgroup is a second row, of the values inside the chosen
   group. The rows below are the records in both, and a new one goes into
   both. The address says which: #/q/<quill>/<screen>/<group>/<subgroup>.

   A field the datamodel marks `secret` is never in a listing; the server
   sends it as null. So a row says it is there — dots — and a tap on the
   eye fetches that one record and shows it, until tapped again or the
   screen is left. The copy button fetches it without showing it. On the
   record sheet it is a password box with the same eye.

   Nothing here knows what it is drawing: a vault and an environment are a
   group and a subgroup, as a customer and a project would be. */

import { api, app, esc, go, heading, icons, nav, replace, store, tabs, toast, wireShell } from "./core.js";
import { installCard } from "./install.js";
import {
  aOr, circle, fieldOf, linkTitles, recordsUrl, sheetHash, spoken, timeLeft, titleField, titleOf, wireAdd,
} from "./kit.js";

export const MASK = "••••••••";

export const secretIcons = {
  show: '<svg width="22" height="22" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.9" stroke-linecap="round" stroke-linejoin="round"><path d="M2 12s3.6-7 10-7 10 7 10 7-3.6 7-10 7S2 12 2 12z"/><circle cx="12" cy="12" r="3"/></svg>',
  hide: '<svg width="22" height="22" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.9" stroke-linecap="round" stroke-linejoin="round"><path d="M10.6 5.1A10 10 0 0 1 12 5c6.4 0 10 7 10 7a17 17 0 0 1-3.1 3.9M6.6 6.6C3.7 8.4 2 12 2 12s3.6 7 10 7a9.7 9.7 0 0 0 5.4-1.6"/><path d="M9.9 9.9a3 3 0 0 0 4.2 4.2M3 3l18 18"/></svg>',
  copy: '<svg width="20" height="20" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.9" stroke-linecap="round" stroke-linejoin="round"><rect x="9" y="9" width="11" height="11" rx="1"/><path d="M5 15V5a1 1 0 0 1 1-1h10"/></svg>',
};

/** Whether a list screen is drawn here rather than by kit.js's plain list. */
export const drawsHere = (model, screen) =>
  !!(screen.group || (screen.subtitle && (fieldOf(model, screen.subtitle) || {}).secret));

/** The record sheet's widget for a secret field: a password box and the eye. */
export function secretWidget(f, value) {
  return `<span class="secret-field"><input type="password" data-field="${esc(f.name)}"` +
    ` aria-label="${esc(f.label)}" value="${esc(value ?? "")}" autocomplete="off" spellcheck="false"` +
    ` placeholder="${f.required ? "" : "hidden"}">` +
    `<button type="button" class="reveal" aria-label="Show ${esc(f.label.toLowerCase())}" aria-pressed="false">${secretIcons.show}</button></span>`;
}

/** The eyes on a sheet: each shows its box, or hides it again. */
export function wireSecretWidgets(root) {
  for (const button of root.querySelectorAll(".secret-field .reveal")) {
    button.addEventListener("click", () => {
      const input = button.parentElement.querySelector("input");
      const shown = input.type === "password";
      input.type = shown ? "text" : "password";
      button.innerHTML = shown ? secretIcons.hide : secretIcons.show;
      button.setAttribute("aria-pressed", String(shown));
    });
  }
}

// -- the levels ------------------------------------------------------------------------
/** What a level offers, as [value, label] pairs, in order. */
async function choicesOf(quill, f, records) {
  if (f.kind === "link") {
    const target = quill.models[f.to];
    if (!target) return [];
    return (await api("GET", recordsUrl(target.id))).map((r) => [r.id, titleOf(target, r)]);
  }
  if (f.kind === "enum") return f.values.map((v, i) => [v, (f.labels && f.labels[i]) || v]);
  const seen = new Set(records.map((r) => r.fields[f.name]).filter((v) => v !== null && v !== undefined && v !== ""));
  return [...seen].sort((a, b) => String(a).localeCompare(String(b))).map((v) => [v, String(v)]);
}

/** The value a level stands on: the one asked for, else the one you had, else the default, else the first. */
function pick(choices, asked, remembered, fallback) {
  const has = (v) => v && choices.some(([c]) => c === v);
  if (asked) return asked;
  if (has(remembered)) return remembered;
  if (has(fallback)) return fallback;
  return choices.length ? choices[0][0] : fallback || null;
}

export async function renderGroupedList(at, arg) {
  const { quill, screen } = at;
  const model = quill.models[screen.model];
  const title = titleField(at, model);
  const tick = screen.tick ? fieldOf(model, screen.tick) : null;
  const under = screen.subtitle ? fieldOf(model, screen.subtitle) : null;
  const levels = [screen.group, screen.subgroup].map((n) => (n ? fieldOf(model, n) : null)).filter(Boolean);
  const asked = String(arg || "").split("/").filter(Boolean);
  const remembered = (() => { try { return JSON.parse(store.get("kit." + at.tab) || "[]"); } catch { return []; } })();

  let all = await api("GET", recordsUrl(model.id));
  const links = under && under.kind === "link" ? await linkTitles(quill, [under]) : {};

  // Each level's choices are the ones inside the level above it.
  const chosen = [];
  const offered = [];
  let inside = all;
  for (const [i, f] of levels.entries()) {
    const choices = await choicesOf(quill, f, inside);
    const value = pick(choices, asked[i], remembered[i], f.default);
    // A value nothing has yet — a vault just named — is still somewhere to stand.
    if (value !== null && !choices.some(([c]) => c === value)) choices.push([value, String(value)]);
    chosen.push(value);
    offered.push(choices);
    inside = inside.filter((r) => value === null || r.fields[f.name] === value);
  }
  try { store.set("kit." + at.tab, JSON.stringify(chosen)); } catch { /* private window */ }
  const here = () => all.filter((r) => levels.every((f, i) => chosen[i] === null || r.fields[f.name] === chosen[i]));
  const hashFor = (values) => at.base + values.map((v) => "/" + encodeURIComponent(v)).join("");
  if (asked.length && asked.some((v, i) => v !== chosen[i])) replace(hashFor(chosen));

  const chipRow = (f, i) => `<div class="chips${i ? " sub" : ""}" data-level="${i}">` +
    offered[i].map(([value, label]) => `<a class="chip${value === chosen[i] ? " active" : ""}" href="${
      esc(hashFor([...chosen.slice(0, i), value]))}">${esc(label)}</a>`).join("") +
    `<button class="chip new-group" type="button" data-level="${i}">+ New ${esc(f.label.toLowerCase())}</button></div>`;
  const where = chosen.filter((v) => v !== null).map((v, i) =>
    levels[i].kind === "link" ? (offered[i].find(([c]) => c === v) || [v, v])[1] : String(v));

  app.innerHTML = nav({ title: screen.label }) + `
    <main>
      ${heading(screen.label, `<button class="compose" aria-label="New ${esc(model.label.toLowerCase())}">${icons.compose}</button>`)}
      ${levels.map(chipRow).join("")}
      <form class="add">${tick ? circle("") : ""}<input placeholder="Add ${esc(aOr(model.label))}${
        where.length ? ` to ${esc(where.join(" · "))}` : ""}" autocapitalize="off" autocomplete="off" spellcheck="false" enterkeyhint="done"></form>
      <div class="install-slot">${installCard()}</div>
      <div class="listing"></div>
    </main>` + tabs(at.tab);
  wireShell();
  const listing = app.querySelector(".listing");
  // What has been revealed, by record id, for as long as this screen is up.
  const revealed = new Map();

  const underText = (r) => {
    if (!under) return "";
    if (under.secret) {
      const value = revealed.get(r.id);
      return value === undefined
        ? `<span class="preview mask" aria-label="Hidden">${MASK}</span>`
        : `<span class="preview shown">${esc(value || "(empty)")}</span>`;
    }
    const said = spoken(under, r.fields[under.name], links[under.name]);
    return said ? `<span class="preview">${esc(said)}</span>` : "";
  };
  const row = (r) => {
    const left = timeLeft(r.expires_at);
    const meta = (left ? `<span class="date expires">${left}</span>` : "") + underText(r);
    const text = `<span class="title">${esc(titleOf(model, r, title))}</span>${meta ? `<span class="meta">${meta}</span>` : ""}`;
    const on = tick && !!r.fields[tick.name];
    const eyes = under && under.secret
      ? `<button class="reveal" data-id="${esc(r.id)}" aria-label="${revealed.has(r.id) ? "Hide" : "Show"} ${esc(under.label.toLowerCase())}" aria-pressed="${revealed.has(r.id)}">${revealed.has(r.id) ? secretIcons.hide : secretIcons.show}</button>` +
        `<button class="copy" data-id="${esc(r.id)}" aria-label="Copy ${esc(under.label.toLowerCase())}">${secretIcons.copy}</button>`
      : "";
    return `<div class="row card grouped${on ? " is-done" : ""}${tick ? "" : " no-tick"}">` +
      (tick ? `<button class="tick" data-id="${esc(r.id)}" aria-label="${esc(tick.label)}" aria-pressed="${on}">${circle(on ? "done" : "")}</button>` : "") +
      `<a class="main" href="${sheetHash(at, model.id, r.id)}">${text}</a>${eyes}</div>`;
  };
  const show = () => {
    const rows = here();
    if (!rows.length) {
      const place = where.length ? ` in ${esc(where.join(" · "))}` : "";
      listing.innerHTML = `<p class="empty mascot"><b>Nothing${place} yet</b>Add ${esc(aOr(model.label))} above, and it will show up here.</p>`;
      return;
    }
    listing.innerHTML = `<div class="group kit-list">${rows.map(row).join("")}</div>`;
    for (const button of listing.querySelectorAll(".tick")) button.addEventListener("click", () => flip(button.dataset.id));
    for (const button of listing.querySelectorAll(".reveal")) button.addEventListener("click", () => reveal(button.dataset.id));
    for (const button of listing.querySelectorAll(".copy")) button.addEventListener("click", () => copy(button.dataset.id));
  };

  const valueOf = async (id) => {
    if (revealed.has(id)) return revealed.get(id);
    const one = await api("GET", recordsUrl(model.id, id));
    return one.fields[under.name] ?? "";
  };
  const reveal = async (id) => {
    if (revealed.has(id)) revealed.delete(id);
    else {
      try { revealed.set(id, await valueOf(id)); } catch (err) { toast(err.message); return; }
    }
    show();
  };
  const copy = async (id) => {
    try {
      await navigator.clipboard.writeText(await valueOf(id));
      toast("Copied");
    } catch (err) { toast(err.message || "Could not copy"); }
  };
  const flip = async (id) => {
    const record = all.find((r) => r.id === id);
    if (!record) return;
    try {
      const changed = await api("PATCH", recordsUrl(model.id, id),
        { fields: { [tick.name]: !record.fields[tick.name] }, rev: record.rev });
      all = all.map((r) => (r.id === id ? changed : r));
    } catch (err) {
      toast(err.status === 409 ? "That changed somewhere else" : err.message);
      all = await api("GET", recordsUrl(model.id));
    }
    show();
  };
  show();

  // A new record goes into the group and subgroup on screen, and opens, so
  // what the add row has no room for — a value, a date — is typed next.
  wireAdd(async (text) => {
    const fields = { [title]: text };
    levels.forEach((f, i) => { if (chosen[i] !== null) fields[f.name] = chosen[i]; });
    const made = await api("POST", recordsUrl(model.id), { fields });
    go(sheetHash(at, model.id, made.id));
  });

  for (const button of app.querySelectorAll(".new-group")) {
    button.addEventListener("click", () => newValue(Number(button.dataset.level)));
  }
  async function newValue(i) {
    const f = levels[i];
    if (f.kind === "enum") { toast(`${f.label} is one of the chips`); return; }
    const text = (prompt(`New ${f.label.toLowerCase()}`) || "").trim();
    if (!text) return;
    let value = text.toLowerCase();
    if (f.kind === "link") {
      const target = quill.models[f.to];
      try {
        value = (await api("POST", recordsUrl(target.id), { fields: { [target.title]: text } })).id;
      } catch (err) { toast(err.message); return; }
    }
    // A name with nothing in it yet: the list is empty there until the first is added.
    go(hashFor([...chosen.slice(0, i), value]));
  }
}
